### PR TITLE
Git Progress

### PR DESCRIPTION
**Pull request by tgymnich**
Added a callback and delegate methods to report the progress of git fetch operations.

### Motivation:

Git fetches currently show no progress indicator which might make some users believe spm is hanging.

### Modifications:

Added delegate methods in `WorkspaceDelegate`and `RepositoryManagerDelegate` to handle reporting of git fetch progress.

### Result:

```
tim@Tims-MacBook-Pro swift/swiftpm (git-progress) $ swift run swift-package update --package-path ~/Desktop/TestPackage
Fetching https://github.com/danielgindi/Charts.git
[3816/34399] Receiving objects 18.70 MiB | 1.49 MiB/s
```

### CODE REVIEW

File: Sources/SourceControl/GitProgress.swift
Comment (mattt) [inline Sources/SourceControl/GitProgress.swift:23]:
```suggestion
    /// The pattern used to match git output. Capture groups are labled from ?<i0> to ?<i19>.
```

Comment (tgymnich) [reply Sources/SourceControl/GitProgress.swift:23]: Good catch 😅 !

**Comment by mattt** [comment]
@tgymnich Your [regex](https://github.com/apple/swift-package-manager/pull/3094/files#diff-c1a15c206a31e2374bf46b9fc825490bf1b0bdf82102b3e7ffe181207258492dR25-R41) is a thing of beauty. Nice work!

@tomerd Seeing as how we're starting to take advantage of more sophisticated Git features to improve performance and UX, is there any appetite for transitioning from `git` plumbing to something like [libgit2](https://libgit2.org) (such as [this wrapper](https://github.com/swiftdocorg/git))?

File: Sources/SourceControl/Repository.swift
Comment (abertelrud) [inline Sources/SourceControl/Repository.swift:148]: This should probably instead call the existing version that doesn't have a progress handler.  Otherwise any existing clients will stop working as soon as this is merged (since they all have `main` dependencies currently).

Comment (abertelrud) [reply Sources/SourceControl/Repository.swift:148]: Something like what the corresponding `Repository` extension does below.

**Comment by abertelrud** [comment]
@tgymnich This is great!  Just a comment about one of the callbacks, which could be made more friendly to existing clients.  What is your sense of how much the exact Git output changes between versions; my impression is "not a lot" — and I would imagine that since they don't provide a parsable format and therefore people parse the existing one, there is incentive to keep it the same — but I don't really know how robust this will be.

Anyway, really looking forward to this one!

**Comment by tomerd** [comment]
> @tomerd Seeing as how we're starting to take advantage of more sophisticated Git features to improve performance and UX, is there any appetite for transitioning from git plumbing to something like libgit2 (such as this wrapper)?

definitely! libgit2 is licensed as "GPLv2 with Linking Exception" so we need to make sure we think about that aspect

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Review by abertelrud — APPROVED** [review]
This looks good to me.  Thanks for your work on this!

This is marked as a Draft, but are there any more issues to address that you know of?

**Comment by tgymnich** [comment]
> @tgymnich Your [regex](https://github.com/apple/swift-package-manager/pull/3094/files#diff-c1a15c206a31e2374bf46b9fc825490bf1b0bdf82102b3e7ffe181207258492dR25-R41) is a thing of beauty. Nice work!

Thank you 😊. While working on this I happened to come across an [amazing article](https://nshipster.com/swift-regular-expressions/) about regex. 😄

**Comment by tgymnich** [comment]
> This is marked as a Draft, but are there any more issues to address that you know of?

@abertelrud I would like to generalize `GitProgress` with a protocol so the delegate methods can be used for other kinds of repositories. I also still need to verify if the git output is the same on ubuntu 16 and 18. Unit testing this would be great but that would require a remote git fetch as far as I know.

**Comment by tgymnich** [comment]
Since fetch operations are now concurrent this PR requires further work.

**Comment by tomerd** [comment]
@tgymnich fyi we are going into a merge freeze for non-critical bug fixes until after the 5.4 branch is created on 1/7 (see https://forums.swift.org/t/swift-5-4-release-process/), so will merge it then

**Comment by tomerd** [comment]
@tgymnich let us know if there is anything we can do to help move this forward, its a fantastic improvement!

**Comment by tgymnich** [comment]
@tomerd I still need to decide how we want to display the progress (e.g. multiple progress bars or the way we handle downloads). Right now I'm busy with exams though, I'll pick this back up next month.

**Comment by neonichu** [comment]
Hi @tgymnich, thanks for this PR. Is there anything we can do to help with bringing this over the finish line?

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Comment by abertelrud** [comment]
At this point it looks as if the errors are related to this PR:

```
/home/buildnode/jenkins/workspace/swift-package-manager-Linux-smoke-test@2/branch-main/swiftpm/Tests/FunctionalTests/MiscellaneousTests.swift:413: error: MiscellaneousTestCase.testReportingErrorFromGitCommand : XCTAssertTrue failed - Error from git was not propogated to process output: /home/buildnode/jenkins/workspace/swift-package-manager-Linux-smoke-test@2/branch-main/tmp/Miscellaneous_MissingDependency.sv3s8n/NonExistantPackage: error: Failed to clone repository /home/buildnode/jenkins/workspace/swift-package-manager-Linux-smoke-test@2/branch-main/tmp/Miscellaneous_MissingDependency.sv3s8n/NonExistantPackage:
```

File: Sources/SourceControl/GitRepository.swift
Comment (tgymnich) [inline Sources/SourceControl/GitRepository.swift:34]: `ProcessResult` does not contain the output if the output is redirected...

Comment (tgymnich) [reply Sources/SourceControl/GitRepository.swift:34]: Some .streamAndCollect option might be one possible solution

File: Sources/Workspace/Workspace.swift
Comment (abertelrud) [inline Sources/Workspace/Workspace.swift:97]: I think a lot of these were removed so that clients need to implement them.  Should these be added as defaults here in this PR?

File: Sources/SourceControl/GitRepository.swift
Comment (abertelrud) [reply Sources/SourceControl/GitRepository.swift:34]: I see — I'll take a closer look and see if what I can suggest here.  Do you know whether this is something that changed through other commits in the meantime, while this PR was up?

Comment (abertelrud) [reply Sources/SourceControl/GitRepository.swift:34]: Having a `.streamAndCollect` does indeed look interesting — I guess in this case you want to stream `stderr` but collect `stdout`.

I almost seems to me as if stream and collect should be booleans, not exclusive choices — with `.none`, `.stream`, and `.collect` as already being options, adding `.streamAndCollect` seems like a lesser choice than just letting each of the outputs have both an optional streaming closure as well as an optional collection container, independently controllable.

Comment (abertelrud) [reply Sources/SourceControl/GitRepository.swift:34]: Actually, in this case `GitRepository.callGit()` could collect the data in the stream closure, and then add it to the `GitCloneError` instead of having a whole `ProcessResult` as the third parameter.

**Comment by abertelrud** [comment]
@tgymnich I have cloned your PR branch and added a workaround for the Git issue here: https://github.com/abertelrud/swift-package-manager/tree/git-progress.  WDYT?  They're not ideal but I think we need TSC changes to clean it up, and I'd rather not hold this back any longer.

There are also some merge conflicts between this PR and main now, and I'm happy to update to main and resolve those.  Is it OK if I push more changes to your branch, or do you want to apply changes (after discussion to see if you agree with them of course)?  Thanks for your patience with this!

**Comment by tgymnich** [comment]
Thank you so much for stepping in. Of course you can push to my branch. Your changes look good to me. Sorry for my slow response, I've got lots to do at the moment.

**Comment by abertelrud** [comment]
> @abertelrud Thank you so much for stepping in. Of course you can push to my branch. Your changes look good to me. Sorry for my slow response, I've got lots to do at the moment.

Thanks!  If anyone should apologize for a slow response, it's me!  :)  Thanks a lot for your original work on this!  This will make a lot of people happy.

I will update your branch and then see if we can get this green.  And follow up with further TSC improvements that might clean up this code.  Thanks!

**Comment by abertelrud** [comment]
Pushed my changes.  Running tests and will rebase on main and fix failures.

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Comment by abertelrud** [comment]
Ok, merged to main and dealt with the resulting conflicts, plus accounted for the repository refactoring that had meanwhile happened on main.

**Comment by abertelrud** [comment]
Part of the refactoring was to follow the pattern of not using default implementations of methods, so I went ahead with that as part of making things build again.

**Comment by abertelrud** [comment]
@tgymnich Please let me know if you're still OK with the changes.  The commits are still separated, and properly attributed to each of us so we can see the history.

Thanks again for your work on this!  I will do automated and some manual testing and then finally get this merged!

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Comment by abertelrud** [comment]
The Linux failure looks unrelated...
```
Building the standard library for: swift-stdlib-linux-x86_64
Running Swift tests for: check-swift-validation-linux-x86_64
ERROR: command terminated with a non-zero exit status 2, aborting
```

**Review by tgymnich — COMMENTED** [review]
LGTM

**Comment by abertelrud** [comment]
@swift-ci please smoke test linux

**Comment by abertelrud** [comment]
Woohoo!  All tests pass, moving to ready-to-merge.  Thanks @tgymnich!

**Comment by abertelrud** [comment]
@neonichu @tomerd Are you fine with this as well?

File: Sources/Commands/SwiftTool.swift
Comment (tomerd) [inline Sources/Commands/SwiftTool.swift:59]: does this needs to be thread safe? if so use `ThreadSafeKeyValueStore`

File: Sources/SourceControl/GitRepository.swift
Comment (tomerd) [inline Sources/SourceControl/GitRepository.swift:967]: does this and all its method need to be public?

Comment (tomerd) [reply Sources/SourceControl/GitRepository.swift:967]: GitFetchProgress -> GitProgressParser

File: Sources/SourceControl/GitRepository.swift
Comment (tomerd) [inline Sources/SourceControl/GitRepository.swift:1098]: this feels out of place here, maybe lift back into GitRepository as a private function?

File: Sources/SourceControl/GitRepository.swift
Comment (tomerd) [inline Sources/SourceControl/GitRepository.swift:1101]: does this work on all platforms? is there not builtin way to do newline delimited split?

File: Sources/SourceControl/Repository.swift
Comment (tomerd) [inline Sources/SourceControl/Repository.swift:63]: progress -> progressHandler

File: Sources/SourceControl/Repository.swift
Comment (tomerd) [inline Sources/SourceControl/Repository.swift:159]: progress -> progressHandler

File: Sources/SourceControl/RepositoryManager.swift
Comment (tomerd) [inline Sources/SourceControl/RepositoryManager.swift:236]: iiuc the queue is for the delegate, not the action? if so, should it be `delegateQueue: DispatchQueue`

File: Sources/SourceControl/RepositoryManager.swift
Comment (tomerd) [inline Sources/SourceControl/RepositoryManager.swift:242]: flip line 241 and 242 to save redundant queue operations

**Review by tomerd — COMMENTED** [review]
very nice! some questions / ideas

File: Sources/Commands/SwiftTool.swift
Comment (tgymnich) [reply Sources/Commands/SwiftTool.swift:59]: `fetchProgress` is only accessed from within the "org.swift.swiftpm.commands.tool-workspace-delegate" queue, this shouldn't need to be thread safe.

File: Sources/SourceControl/GitRepository.swift
Comment (tgymnich) [reply Sources/SourceControl/GitRepository.swift:1101]: `String.isNewline` should be just what we need here.

File: Sources/SourceControl/RepositoryManager.swift
Comment (tgymnich) [reply Sources/SourceControl/RepositoryManager.swift]: done 7983807

File: Sources/SourceControl/GitRepository.swift
Comment (tgymnich) [reply Sources/SourceControl/GitRepository.swift]: done 5a5ebc6

File: Sources/SourceControl/Repository.swift
Comment (tgymnich) [reply Sources/SourceControl/Repository.swift:159]: done 6467569

File: Sources/SourceControl/Repository.swift
Comment (tgymnich) [reply Sources/SourceControl/Repository.swift]: done 6467569

File: Sources/SourceControl/RepositoryManager.swift
Comment (tgymnich) [reply Sources/SourceControl/RepositoryManager.swift]: done cdac447

File: Sources/SourceControl/GitRepository.swift
Comment (tgymnich) [reply Sources/SourceControl/GitRepository.swift]: done e98d6b5

File: Sources/SourceControl/GitRepository.swift
Comment (tgymnich) [reply Sources/SourceControl/GitRepository.swift:1098]: `gitFetchStatusFilter` is called from both `GitRepositoryProvider` and  `GitRepository` so we could put it at the end of GitRepository.swift` as a `fileprivate func`. Maybe like this be29d1.

**Review by abertelrud — APPROVED** [review]

**Comment by abertelrud** [comment]
Thanks for the additional changes!

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Review by neonichu — APPROVED** [review]

**Review by tomerd — APPROVED** [review]

**Comment by tomerd** [comment]
thanks @tgymnich looking forward to this getting merged and in production!

File: Sources/Commands/SwiftTool.swift
Comment (abertelrud) [reply Sources/Commands/SwiftTool.swift:59]: Indeed, it seems to be exactly along the same lines as `downloadProgress` above.

**Pull request merged**
